homeworks/solutions/krasnor/1-intro/src/BasicOBJImporter.js

```diff
@@ -0,0 +1,196 @@
+import {FileLoader} from 'three/src/loaders/FileLoader';
+import {
+    BufferGeometry, Vector2, Vector3, Face3, Geometry, Group, Mesh,
+} from 'three'
+
+
+/**
+ * A very basic implementation of an .obj importer.
+ *
+ * This importer  only supports triangulated faces. No Quad Faces
+ * Only Object Name, Mesh and UV data is imported.
+ *
+ * Other features such as material libraries, groups, Free-form curve/surface, etc. are not supported by this loader.
+ */
+export default class BasicOBJImporter {
+    constructor() {
+        // nothing to do
+    }
+
+    /**
+     *
+     * @param {String} assetPath is a String
+     * @returns {Group} A Group containing all parsed meshes.
+     */
+    async importObj(assetPath) {
+        let scope = this;
+        return new Promise(function (resolve, reject) {
+            let tmp_fileLoader = new FileLoader();
+            // tmp_fileLoader.setResponseType('arraybuffer');
+            // tmp_fileLoader.setResponseType('blob');
+
+            //load the obj file and parse the result
+            //console.log('## 1 call FileLoader');
+            tmp_fileLoader.load(
+                // resource URL
+                assetPath,
+                // onLoad callback
+                function (data) {
+                    // loading finished - do parsing
+                    try{
+                        let parsed_mesh = scope.parse(data);
+                        //console.log('## 3 parsed mesh')
+                        resolve(parsed_mesh);
+                    }catch (e) {
+                        reject('An error happened: ' + e);
+                    }
+                },
+                // onProgress callback
+                function (xhr) {
+                    //console.log( (xhr.loaded / xhr.total * 100) + '% loaded' );
+                },
+                // onError callback
+                function (err) {
+                    reject('An error happened: ' + err);
+                }
+            );
+            //console.log('## 2 after FileLoader call')
+        });
+
+    }
+
+    /**
+     *
+     * @param {ArrayBuffer} assetPath is a String
+     */
+    parse(content) {
+        //  console.log( content );
+
+        // iterate over each line and process it
+        // TODO research, if ArrayBuffer, or Blob could be used for more efficient reading
```

Comment:
Good point. This is actually why we do not really like `.obj` that much because from the specification it does not give us any hint to seek that information, such as from which location we can load face data. Ideally, we would love to pre-allocate the array, and further, we even want do some parallel reading speedup. unfortunately, that information is not available until we read the whole file :(